feat: kick off the Python course (#1023)
apify/apify-docs: sources/academy/webscraping/scraping_basics_python/07_extracting_data.md

---
title: Extracting data from HTML with Python
sidebar_label: Extracting data from HTML
description: Lesson about building a Python application for watching prices. Using string manipulation to extract and clean data scraped from the product listing page.
sidebar_position: 7
slug: /scraping-basics-python/extracting-data
---

**In this lesson we'll finish extracting product data from the downloaded HTML. With help of basic string manipulation we'll focus on cleaning and correctly representing the product price.**

---

Locating the right HTML elements is the first step of a successful data extraction, so it's no surprise that we're already close to having the data in the correct form. The last bit that still requires our attention is the price:

```text
$ python main.py
JBL Flip 4 Waterproof Portable Bluetooth Speaker $74.95
Sony XBR-950G BRAVIA 4K HDR Ultra HD TV From $1,398.00
Sony SACS9 10" Active Subwoofer $158.00
Sony PS-HX500 Hi-Res USB Turntable $398.00
...
```

Let's summarize what stands in our way if we want to have it in our Python program as a number:

- A dollar sign preceeds the number,

> - the number contains decimal commas so it better reads by humans,

Reviewer: ```suggestion
- the number contains decimal commas for better human readability,
```

if price_text.startswith("From "):
    min_price = price_text.removeprefix("From ")
    max_price = None
else:
    min_price = price_text
    max_price = min_price
```

We're using Python's built-in string methods:

- `.startswith()`, a method for [checking the beginning of a string](https://docs.python.org/3/library/stdtypes.html#str.startswith).
- `.removeprefix()`, a method for [removing something from the beginning of a string](https://docs.python.org/3/library/stdtypes.html#str.removeprefix).

The whole program now looks like this:

```py
import httpx
from bs4 import BeautifulSoup

url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
response = httpx.get(url)
response.raise_for_status()

html_code = response.text
soup = BeautifulSoup(html_code, "html.parser")

for product in soup.select(".product-item"):
    title = product.select_one(".product-item__title").text

    price_text = product.select_one(".price").contents[-1]
    if price_text.startswith("From "):
        min_price = price_text.removeprefix("From ")
        max_price = None
    else:
        min_price = price_text
        max_price = min_price

>     print(title, min_price, max_price)

Reviewer: I don't think this is how we would represent it in a real dataset, but perhaps it doesn't matter.

The thing is, this makes it hard to know whether the price is real, or an interval. As the consumer of this dataset you have to make additional comparison `if min_price == max_price` to see whether you have an actual price. Might be better to save real prices to `price` and use `min_price` `max_price` only for intervals.

But again, not a strong opinion.

Reviewer: I would also mention why there's an interval (because the product has variants with different prices), and that the real prices can be retrieved from the product detail pages, which we will scrape later (we will, right? 😄)

Author: I'd love to make it more plausible as a real-world case, but I'm no data analyst, so it's hard for me to assess how this would be best approached. I came up with an interval as something that can capture the nature of the data, but you're right about the additional comparison. Would something like this make sense?

- If there's exact price, both `price` and `min_price` are set to the number
- If there's just minimal price, `price` is set to `None`, and `min_price` is set to the number
- There's no `max_price`

Once we scrape the product pages, we'd just fill in the missing `price` 🤔

Author: Addressed in a39846d1

```text
https://www.theguardian.com/sport/formulaone
```

Your program should print something like the following. Note the dates at the end of each line:

```text
Wolff confident Mercedes are heading to front of grid after Canada improvement 2024-06-10
Frustrated Lando Norris blames McLaren team for missed chance 2024-06-09
Max Verstappen wins Canadian Grand Prix: F1 – as it happened 2024-06-09
...
```

Hints:

- HTML's `<time>` tag can have an attribute `datetime`, which [contains data in a machine-readable format](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time), such as the ISO 8601.
- Beautiful Soup gives you [access to attributes as if they were dictionary keys](https://beautiful-soup-4.readthedocs.io/en/latest/#attributes).
- In Python you can create `datetime` objects using `datetime.fromisoformat()`, a [built-in method for parsing ISO 8601 strings](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat).
- To get just the date part, you can call `.date()` on any `datetime` object.

<details>
  <summary>Solution</summary>

  ```py

>   import httpx
>   from bs4 import BeautifulSoup

Reviewer 2: Missing import :slightly_smiling_face:.

```suggestion
  import httpx
  from datetime import datetime
  from bs4 import BeautifulSoup
```
